core/ir/synthetic/encode.c

```diff
@@ -0,0 +1,129 @@
+ * DAMAGE.
+ */
+
+/* encode.c -- an encoder for DR synthetic IR */
+
+#include "encode.h"
+
+#include "../globals.h"
+#include "encode_api.h"
+#include "encoding_common.h"
+#include "instr_api.h"
+#include "opnd_api.h"
+
+/* Encodes DR instruction representation \p instr into raw bytes \p encoded_instr.
+ * Returns next instruction's pc.
+ * The encoding scheme followed is described in core/ir/synthetic/encoding_common.h.
+ */
+byte *
+encode_to_synth(dcontext_t *dcontext, instr_t *instr, byte *encoded_instr)
+{
+    /* Use a local uint variable for easier setting of category, eflags, #src, and #dst
+     * values.
+     */
+    uint encoding_header = 0;
+
+    /* Encode number of register destination operands (i.e., written registers).
+     */
+    uint num_dsts = (uint)instr_num_dsts(instr);
+    encoding_header |= num_dsts;
+
+    /* Encode number of register source operands (i.e., read registers).
+     */
+    uint num_srcs = (uint)instr_num_srcs(instr);
+    encoding_header |= (num_srcs << SRC_OPND_SHIFT);
+
+    /* Encode arithmetic flags.
+     */
+    uint eflags_instr = instr_get_arith_flags(instr, DR_QUERY_DEFAULT);
+    uint eflags = 0;
+    if (TESTANY(EFLAGS_WRITE_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_WRITES_ARITH;
+    if (TESTANY(EFLAGS_READ_ARITH, eflags_instr))
+        eflags |= SYNTHETIC_INSTR_READS_ARITH;
+    encoding_header |= (eflags << FLAGS_SHIFT);
+
+    /* Encode category as synthetic opcode.
+     */
+    uint category = instr_get_category(instr);
+    encoding_header |= (category << CATEGORY_SHIFT);
+
+    /* Copy header encoding back into encoded_instr output.
+     */
+    *((uint *)&encoded_instr[0]) = encoding_header;
+
+    /* Encode register destination operands, if present.
+     */
+    opnd_size_t max_opnd_size = OPSZ_NA;
+    for (uint dst_index = 0; dst_index < num_dsts; ++dst_index) {
+        opnd_t dst_opnd = instr_get_dst(instr, dst_index);
+        max_opnd_size = opnd_get_size(dst_opnd);
+        uint num_regs_used_by_opnd = (uint)opnd_num_regs_used(dst_opnd);
+        for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+            reg_id_t reg = opnd_get_reg_used(dst_opnd, opnd_index);
+            encoded_instr[dst_index + OPND_INDEX] = (byte)reg;
+        }
+    }
+
+    /* Encode register source operands, if present.
+     */
+    for (uint src_index = 0; src_index < num_srcs; ++src_index) {
+        opnd_t src_opnd = instr_get_src(instr, src_index);
+        max_opnd_size = opnd_get_size(src_opnd);
+        uint num_regs_used_by_opnd = (uint)opnd_num_regs_used(src_opnd);
+        for (uint opnd_index = 0; opnd_index < num_regs_used_by_opnd; ++opnd_index) {
+            reg_id_t reg = opnd_get_reg_used(src_opnd, opnd_index);
+            encoded_instr[src_index + OPND_INDEX + num_dsts] = (byte)reg;
+        }
+    }
+
+    /* Encode largest register size, if there is at least one operand.
```

Comment:
See other comments on this not being limited to registers.

Reply:
Done.